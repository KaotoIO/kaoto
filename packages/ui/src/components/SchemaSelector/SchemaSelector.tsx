import { FunctionComponent, PropsWithChildren, useCallback, useEffect, useState } from 'react';
import { Schema } from '../../models';
import { Catalog, ITile } from '../Catalog';

const SCHEMA_INDEX_KEY = 'Schema';

interface IconResolverProps {
  schemas: Schema[];
  onClick?: (schema: Schema) => void;
}

export const SchemaSelector: FunctionComponent<PropsWithChildren<IconResolverProps>> = (props) => {
  const [tiles, setTiles] = useState<Record<string, ITile[]>>({});

  useEffect(() => {
    const tiles: ITile[] = props.schemas.map((schema) => {
      return {
        name: schema.name,
        title: schema.name,
        description: '',
        tags: [schema.version],
        type: SCHEMA_INDEX_KEY,
        rawObject: schema.schema,
      };
    });

    setTiles({ [SCHEMA_INDEX_KEY]: tiles });
  }, [props.schemas]);

  const onTileClick = useCallback(
    (tile: ITile) => {
      props.onClick?.(tile.rawObject as Schema);
    },
    [props],
  );

  return <Catalog tiles={tiles} onTileClick={onTileClick} />;
};
