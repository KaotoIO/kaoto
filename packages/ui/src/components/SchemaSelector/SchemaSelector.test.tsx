import { fireEvent, render } from '@testing-library/react';
import { SchemaSelector } from './SchemaSelector';
import userSchemaJson from '../../stubs/user-schema.json';

describe('SchemaSelector', () => {
  it('should render correctly', () => {
    const props = {
      schemas: [
        {
          name: 'name',
          version: 'version',
          schema: userSchemaJson,
        },
      ],
    };

    const { container } = render(<SchemaSelector {...props} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call onClick when a tile is clicked', () => {
    const props = {
      schemas: [
        {
          name: 'name',
          version: 'version',
          schema: userSchemaJson,
        },
      ],
      onClick: jest.fn(),
    };

    const { getByText } = render(<SchemaSelector {...props} />);

    fireEvent.click(getByText('name'));

    expect(props.onClick).toHaveBeenCalledWith(props.schemas[0].schema);
  });
});
