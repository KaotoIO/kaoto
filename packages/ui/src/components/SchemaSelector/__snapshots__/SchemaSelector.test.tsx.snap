// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaSelector should render correctly 1`] = `
<form
  class="pf-v5-c-form catalog__filter"
  novalidate=""
>
  <div
    class="pf-v5-l-grid pf-m-all-6-col-on-md pf-m-gutter"
  >
    <div
      class="pf-v5-l-grid__item"
    >
      <div
        class="pf-v5-c-form__group"
      >
        <div
          class="pf-v5-c-form__group-label"
        >
          <label
            class="pf-v5-c-form__label"
            for="search-term"
          >
            <span
              class="pf-v5-c-form__label-text"
            >
              Search term
            </span>
          </label>
           
        </div>
        <div
          class="pf-v5-c-form__group-control"
        >
          <div
            class="pf-v5-c-text-input-group"
          >
            <div
              class="pf-v5-c-text-input-group__main pf-m-icon"
            >
              <span
                class="pf-v5-c-text-input-group__text"
              >
                <span
                  class="pf-v5-c-text-input-group__icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                    />
                  </svg>
                </span>
                <input
                  aria-label="Find by name or tag"
                  class="pf-v5-c-text-input-group__text-input"
                  placeholder="Find by name or tag"
                  type="text"
                  value=""
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-l-grid__item pf-v5-u-text-align-right"
    >
      <div
        class="pf-v5-c-form__group"
      >
        <div
          class="pf-v5-c-form__group-label"
        >
          <label
            class="pf-v5-c-form__label"
            for="element-type"
          >
            <span
              class="pf-v5-c-form__label-text"
            >
              Type
            </span>
          </label>
           
        </div>
        <div
          class="pf-v5-c-form__group-control"
        >
          <div
            aria-label="Select element type"
            class="pf-v5-c-toggle-group"
            role="group"
          >
            <div
              class="pf-v5-c-toggle-group__item"
              data-testid="Schema-catalog-tab"
            >
              <button
                aria-pressed="true"
                class="pf-v5-c-toggle-group__button pf-m-selected"
                id="toggle-group-button-Schema"
                type="button"
              >
                <span
                  class="pf-v5-c-toggle-group__text"
                >
                  Schema
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
